Add support for Ender 3 S1 Pro/Plus stock touch-screen

strnk opened the pull request:
### Description

 This is a WIP development to support the stock firmware of the DWIN touch-screen shipped with Creality Ender 3 S1 Plus and Ender 3 S1 Pro printers.

 I'm mainly looking for feedback regarding the implementation as of now. I used the _DGUS reloaded_ source as a reference, but it brings a lot of redundancy between my E3S1pro code, DGUS reloaded, and DGUS classic. Most the differences are enums of the different screen names, and adresses of the controls and displays inside the touch-screen, along with the screen firmware specific handling and so on. Most notably the command interface to the display (`DGUS_Display`) should be common to all supported DWIN screen which use the DGUS commands over serial.

 I am still struggling to support ADVANCED_PAUSE_FEATURE. I drafted a hackish way of displaying custom messages and option selection using the SD card screen, but it's limited to emulate user confirmation / LCD knob click via ExtUI::UserConfirmRequired. I'm still stuck on supporting the "purge more / stop purging" step through ExtUI. Because of that I left out the runout sensor management for now since I'd like the nozzle parking feature and so on.

 I tried to stick to the ExtUI interface, but could not find a way to move multiple axis at once. Moving between diagonal tramming point is done by moving X then Y which seems sluggish. Not sure if it's bad taste or not to extend the ExtUI interface for these kind of things.

 Last of all, I included various hacks for limitations which make no sense on the original creality firmware, but they are configurable and the stock behavior can be used if needed:
 - the original sd card interface use 4 screens numbered 1/4 to 4/4, each displaying 5 files. I extended that but the screen will show 4/4 for every set of 5 files beyond page 4 as it's a fixed image and not a text control.
 - also added the possibility to traverse directories in the sd card interface
 - the stock touch-screen firmware only contains a few text controls, mainly for the print status, the SD card filenames and the about page. Although the DGUS screen supports scrolling controls, all of those are fixed, so I handled the scrolling from software to provide a bit more flexibility.

### Requirements

Requires an Ender 3 S1 Plus or Pro printer using the stock touch-screen (no knob).

### Benefits

Newer releases of Marlin can be used with the stock S1 Plus/Pro touchscreen, since the original firmware is based on 2.0.x

### Configurations

I uploaded a [test firmware for the S1 Plus printer](https://github.com/strnk/Marlin/releases/tag/ender3s1pro-2.1.x-60921e9) on my repo for testing. The configurations used are [also available](https://github.com/strnk/Marlin_Configurations/tree/import-2.1.x-ender3s1plus/config/examples/Creality/Ender-3%20S1%20Plus).

strnk committed:
Ender 3 S1 Pro/Plus stock touch-screen firmware ExtUI

thinkyhead committed:
strip whitespace, update (c)

thinkyhead commented on Marlin/src/lcd/extui/dgus_e3s1pro/DGUSDisplay.h:
```diff
@@ -0,0 +1,185 @@
+};
+
+template<> inline uint16_t DGUSDisplay::SwapBytes(const uint16_t value) {
+  return ((value << 8) | (value >> 8));
```
My guess is that this needs to be made contingent on the endianness of the target. We have no guarantee that all targets for this display will have the same endianness.

strnk replied:
Indeed. I tried to fix that in [2d2dad7](https://github.com/MarlinFirmware/Marlin/pull/25905/commits/2d2dad772e2db38261e9f5a732c8f69c525d29ea) by resolving byte swaps at compile time. 

I could not use <machine/endian.h> as it does not seems to be available on all platforms.

thinkyhead replied:
Looks like we have a nice simple endian header now.

strnk replied:
Note that although it should work fine on most/all platforms it's not strictly speaking C++ compliant since type punning is undefined behavior. I have yet to see a compiler getting angry at that though.

The generic swap() template I took from the DGUS reloaded code yields the same assembly as GCC's builtins __builtin_bswap16/32/64 so I haven't looked into optimizing it further.

thinkyhead committed:
remove tabs, fix formatting

thinkyhead committed:
Option DGUS_LCD_UI E3S1PRO

thinkyhead committed:
lowercase method names

strnk committed:
lowercase method names

strnk committed:
fix print button behavior when no file is selected

strnk committed:
missing define for DGUS_UI_IS(E3S1PRO)

strnk committed:
lowercase method names

strnk committed:
missing nil END member

strnk committed:
make data exchange with DGUS endianness agnostic

strnk committed:
fix PID setting menu behavior depending on PIDTEMP and PIDTEMPBED configuration

strnk committed:
fix bed PID D parameter fixed point format

thinkyhead committed:
misc. cleanup

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25905

thinkyhead committed:
more cleanup

thinkyhead commented:
I think this is about done as far as the screen implementation itself. I don't feel like getting into the convergence between this and the rest of the DGUS screens, but later on it might be a part of some consolidation. The ExtUI stuff seems good. I may add a CI test for this after I update the example configurations with the new screen. We can then add example configurations for the Ender-3 S1 Pro/Plus to the Configurations repo. Is it now possible to build this and have everything work as it should on the Ender-3 S1 Pro/Plus? If so then we can merge this and the example configs for those machines quite soon!

strnk committed:
fix compilation errors

strnk commented:
I still need to test the filament runout since it's a stock feature of the S1 plus. The laser cutter support can be a future PR.

strnk committed:
add filament runout checks
fix filament extrude/retract menu
fix heater temperature control from filament menu

strnk committed:
fix runout debug message format to show binary rather than '0'/'1' ASCII integers

strnk committed:
debug messages cleanup

strnk committed:
add sanity checks for the Ender-3 S1 plus/pro touchscreen features

strnk committed:
add power loss recovery handlers

strnk committed:
fix home/print buttons translated text while printing

strnk committed:
use configuration.h PREHEAT_1/2 presets as PLA/ABS default presets values

strnk committed:
disable dgus debug messages

strnk commented:
I've been able to test the filament runout and power-loss  recovery features so this is mostly done for me. I'm using this on my own printer so this is print-ready :)

FYI I've submitted a PR for the configurations of the Ender 3 S1 Plus and S1 Pro printers to match this touch-screen implementation (MarlinFirmware/Configurations#946), and also compiled [firmwares for both printers](https://github.com/strnk/Marlin/releases/tag/ender3s1pro-2.1.x-70f195e) using these configurations.

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25905

thinkyhead committed:
Config updates. Use LCD_LANGUAGE

strnk committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-ender3s1pro-touchscreen

strnk committed:
fix endianness conversion methods to allow compile-time constexpr initialization

strnk committed:
missing int test in is_integral<>

strnk committed:
fix the remaining print duration counter on home screen

strnk committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x-ender3s1pro-touchscreen

strnk committed:
Merge branch 'MarlinFirmware:bugfix-2.1.x' into bugfix-2.1.x-ender3s1pro-touchscreen

strnk committed:
allow FILAMENT_RUNOUT_SCRIPT to not be defined if unused

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25905

thinkyhead committed:
formatting

thinkyhead committed:
simplify sdCardPrepare

thinkyhead committed:
Just E3S1PRO here

thinkyhead committed:
DGUS_Screen => DGUS_ScreenID

fnsign commented:
Im receiving following warning when compiling on my Ender 3 Pro w/ SKR mini v1.2:
Marlin/src/feature/runout.cpp: In function 'void event_filament_runout(uint8_t)':
Marlin/src/feature/runout.cpp:104:14: warning: unused variable 'park_or_pause' [-Wunused-variable]
  104 |   const bool park_or_pause = (false
